contracts/Smt.metasol

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+
+library SMT{{hashname}} {
+    // because maxElements = 2**depth, so 256 will overflow
+    uint8 constant MAX_DEPTH = 255;
+    bytes32 constant internal EMPTY_SUBTREE = 0x00;
+
+    struct MerkleTree {
+        uint8 depth;
+        mapping(uint256 => bytes32) cacheEmptyValues;
+        // tree level => index inside level => element hash
+        mapping(uint256 => mapping(uint256 => bytes32)) tree;
+        mapping(uint256 => bytes) elementData;
+    }
+
+    using SMT{{hashname}} for MerkleTree;
+
+    function initialize(MerkleTree storage tree, uint8 _depth) internal {
+        require(_depth > 0, "Depth must be non-zero");
+        tree.cacheEmptyValues[0] = {{hash}}("");
+        // current depth == 0
+        tree._calculateEmptyLeafHash(_depth);
+        tree.depth = _depth;
+    }
+
+    function totalElements(MerkleTree storage tree) internal view returns (uint256) {
+        // 1<<depth == 2**depth
+        return 1<<tree.depth;
+    }
+
+    function increaseDepth(MerkleTree storage tree, uint8 depthDifference) internal {
+        require(depthDifference > 0, "depthDifference must be non-zero");
+        uint8 oldDepth = tree.depth;
+        require(depthDifference <= MAX_DEPTH - oldDepth, "Overflow protection");
+        uint8 newDepth = oldDepth + depthDifference;
+
+        tree._calculateEmptyLeafHash(newDepth);
+
+        for (uint8 level = oldDepth+1; level <= newDepth; level += 1) {
+            tree._updateNode(level, 0);
+        }
+        tree.depth = newDepth;
+    }
+
+    function decreaseDepth(MerkleTree storage tree, uint8 depthDifference) internal {
+        require(depthDifference > 0, "depthDifference must be non-zero");
+        uint8 oldDepth = tree.depth;
+        require(depthDifference < oldDepth, "Underflow protection");
+        uint8 newDepth = oldDepth - depthDifference;
+
+        for (uint8 level = newDepth; level < oldDepth; level += 1) {
+            require(tree.tree[level][1] == EMPTY_SUBTREE, "Subtree must be empty");
+            // If we will delete this:
+            // delete tree[level+1][0];
+            // delete cacheEmptyValues[level+1];
+            // refund will be smaller than gas spent :(
+        }
+        tree.depth = newDepth;
+    }
+
+    function getRoot(MerkleTree storage tree) internal view returns (bytes32) {
+        if (tree.tree[tree.depth][0] == EMPTY_SUBTREE) {
+            return tree.cacheEmptyValues[tree.depth];
+        }
+        return tree.tree[tree.depth][0];
+    }
+
+    function _calculateEmptyLeafHash(MerkleTree storage tree, uint8 toLevel) internal {
+        uint8 currentDepth = tree.depth;
+        bytes32 prev = tree.cacheEmptyValues[currentDepth];
+
+        for (uint8 level = currentDepth+1; level <= toLevel; level += 1) {
```

Comment:
I think we should save a number of default nodes in `cacheEmptyValues` that we calculate already. This way we will not recalculate the already calculated values.

Reply:
Do you mean what's proposed here: https://github.com/andreysobol/sparse_merkle_sol/pull/1#discussion_r720303552 ?
If so, then I am for it, but please see its subsequent reply.

Reply:
I believe we can think of a solution that will be both more elegant and working :ok_hand:

Reply:
We assume here that `decreaseDepth` function will not be called ever, it is provided just in case. On the other hand, `increaseDepth` will definitely be called. So trying to optimize this further either works worse (in terms of gas) for our use-case or will cost us more during deployment. Either way, we believe at this point this is overengineering and we should leave it as it is, unless there are any more objections.